// @google/clasp/src/dotfile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Manages dotfiles. There are 2 types of dotfiles:
 *
 * Global clasp auth settings:
 * - File: ~/.clasp.json
 * - Default credentials for clasp projects.
 *
 * Local clasp auth settings:
 * - File: .clasp.json
 * - Requires `clasp login --creds creds.json`
 *
 * This should be the only file that uses DOTFILE.
 */
var fs = require("fs");
var os = require("os");
var path = require("path");
var dotf = require('dotf');
var read = require('read-file');
var findParentDir = require('find-parent-dir');
var splitLines = require('split-lines');
// TEMP CIRCULAR DEPS, TODO REMOVE
// import { PROJECT_NAME } from './utils';
var PROJECT_NAME = 'clasp';
// Dotfile names
exports.DOT = {
    /**
     * This dotfile stores information about ignoring files on `push`. Like .gitignore.
     */
    IGNORE: {
        DIR: '~',
        NAME: PROJECT_NAME + "ignore",
        PATH: "." + PROJECT_NAME + "ignore",
    },
    /**
     * This dotfile saves clasp project information, local to project directory.
     */
    PROJECT: {
        DIR: path.join('.', '/'),
        NAME: PROJECT_NAME + ".json",
        PATH: "." + PROJECT_NAME + ".json",
    },
    /**
     * This dotfile saves auth information. Should never be committed.
     * There are 2 types: personal & global:
     * - Global: In the $HOME directory.
     * - Personal: In the local directory.
     * @see {ClaspToken}
     */
    RC: {
        DIR: '~',
        LOCAL_DIR: './',
        NAME: PROJECT_NAME + "rc.json",
        LOCAL_PATH: "." + PROJECT_NAME + "rc.json",
        PATH: path.join('~', "." + PROJECT_NAME + "rc.json"),
        ABSOLUTE_PATH: path.join(os.homedir(), "." + PROJECT_NAME + "rc.json"),
        ABSOLUTE_LOCAL_PATH: path.join('.', "." + PROJECT_NAME + "rc.json"),
    },
};
// Methods for retrieving dotfiles.
exports.DOTFILE = {
    /**
     * Reads DOT.IGNORE.PATH to get a glob pattern of ignored paths.
     * @return {Promise<string[]>} A list of file glob patterns
     */
    IGNORE: function () {
        var projectDirectory = findParentDir.sync(process.cwd(), exports.DOT.PROJECT.PATH) || exports.DOT.PROJECT.DIR;
        return new Promise(function (res, rej) {
            if (fs.existsSync(path.join(projectDirectory, exports.DOT.IGNORE.PATH))) {
                var buffer = read.sync(exports.DOT.IGNORE.PATH, 'utf8');
                res(splitLines(buffer).filter(function (name) { return name; }));
            }
            else {
                res([]);
            }
        });
    },
    /**
     * Gets the closest DOT.PROJECT.NAME in the parent directory of the directory
     * that the command was run in.
     * @return {dotf} A dotf with that dotfile. Null if there is no file
     */
    PROJECT: function () {
        var projectDirectory = findParentDir.sync(process.cwd(), exports.DOT.PROJECT.PATH) || exports.DOT.PROJECT.DIR;
        return dotf(projectDirectory, exports.DOT.PROJECT.NAME);
    },
    // Stores {ClaspCredentials}
    RC: dotf(exports.DOT.RC.DIR, exports.DOT.RC.NAME),
    // Stores {ClaspCredentials}
    RC_LOCAL: function () {
        var localDirectory = findParentDir.sync(process.cwd(), exports.DOT.PROJECT.PATH) || exports.DOT.RC.LOCAL_DIR;
        return dotf(localDirectory, exports.DOT.RC.NAME);
    },
};
